Export clips v2

JacksonRG opened the pull request:
# Summary
Create a popup window to export clips as their own video files on right click in the files window. If a file, not a clip is selected, copy it. All new files are placed in a folder of the user's choosing.

Here's an example of the ui.
![ExportClipsLabelOnly](https://user-images.githubusercontent.com/42394129/146100681-33c49e04-7d8d-4e8f-9d1e-c76dc9425616.png)
Please note that the green outline, and lack of title bar are due to my DE.

# Use Case
A very common use-case for OpenShot is splitting many small videos from a larger one. For example, when a home video tape collection is digitized, and many different moments are combined into a single video file. OpenShot has a super easy way to split videos into clips (`Project Files->Split Clip...`). However, these are only "clips" within OpenShot, and many users want an easy way to then export these "clips" as actual individual video files.  This PR is aimed at addressing this work-flow.

# Goals
- [x] If the user selects files, copy them to the destination directory
- [x] Export using crf (22) to prevent very large file exports.
    - Possibly add a setting for this.
- [x] Add a progress bar
- [x] Add documentation for this UI
- [x] If file name already exists, don't export (incredibly unlikely that a file is naturally named `file_name [15 to 20].mp4`
- [x] Add copyright info

Commit:
Added Export Clips action to context file menus

Commit:
Created a widget to suggest sub-directories to export to.

Commit:
Popup a dialog for exporting clips.

Prompt user to pick a destination folder.
Start exporting clips when export is clicked.
Update the user when the job is done.

JacksonRG commented on src/windows/main_window.py:
```diff
@@ -1955,6 +1955,20 @@ def actionFile_Properties_trigger(self):
+        exp.setPath(get_app().project.get("export_path"))
+        # f = self.selected_files()
+        # exp = clipExportWindow()
+        # exp.setPath(get_app().project.current_filepath)
```
Just noticed this  commented code. Removing them in a later commit.

Commit:
Cleanup, and codacy fixes

Removed comments in main_window.py
Removed unnecessary imports
Made a couple methods into functions
Correct a few typos

Commit:
Add Documentation for exporting clips

jonoomph commented on src/classes/filePicker.py:
```diff
@@ -0,0 +1,86 @@
+from PyQt5.QtWidgets import QPushButton, QFileDialog, QCompleter, QLineEdit, QLabel
```
Copyright / License info is missing

jonoomph commented:
@JacksonRG Can you add a screenshot of this UI to the description of this PR? Thx

jonoomph commented on src/windows/exportClips.py:
```diff
@@ -0,0 +1,142 @@
+from PyQt5.QtWidgets import QVBoxLayout, QPushButton, QDialog, QLabel
```
Ditto... missing copyright/license

jonoomph commented on src/windows/main_window.py:
```diff
@@ -1955,6 +1955,16 @@ def actionFile_Properties_trigger(self):
+        try:
+            exp.exec_()
+        except:
+            print("Error occured")
```
Replace print with `log.error()`

Commit:
UI file and progress bar

- added a ui file.
- added a progress bar
- moved buttons into a button box, so they look much better.

Commit:
Adopt OpenShot naming convention

Took a cue from export.py, and used camelCase for
function names and snake_case for variables.

Commit:
Naming convention for filePicker.py

Commit:
Added license comments to new files.

Commit:
Added support for files

When filese are selected, they will be given a suffix
to match clips being exported, and copied to the
specified directory.

jonoomph commented:
Here is the format of our normal Export dialog (the footer). I would suggest we copy the same style (alignment, button placement, etc...):
![Export-Dialog-Footer](https://user-images.githubusercontent.com/5551883/145621875-17c7a75a-cf75-40a1-b890-11ba0344de6a.png)

jonoomph commented:
Also, no need for a "Close" button, since we have a "Cancel" button

jonoomph commented on src/classes/filePicker.py:
```diff
@@ -0,0 +1,115 @@
+"""
+ @file filePicker.py
```
I'm confused by this filePicker.py approach? Is there a reason we can't use a general folder picker: `QFileDialog.getOpenFileUrls`, or something similar? Do we need to create our very own `QFileDialog`?

jonoomph replied:
I'm worried that instead of a 1 or 2 line solution, we've introduced a bunch of unneeded complexity to maintain

JacksonRG replied:
It started with me thinking there was an existing widget for typing in file paths, and suggesting sub-directories.

When I didn't find one, I figured we could make one to plug into other widgets, like properties view, or exporter. A quick, consistent way to find files without popping up a dialog.

JacksonRG replied:
I see that it _is_ scope creep for this PR. So if you like, I can:

1) Move this widget to a new branch for sometime down the line.
2) Move this code (minus the auto-complete) into exportClips.py

jonoomph commented on src/windows/exportClips.py:
```diff
@@ -0,0 +1,236 @@
+"""
```
This filename should be all lower-case, with _ between words (like the other window classes)

jonoomph commented on src/windows/exportClips.py:
```diff
@@ -0,0 +1,236 @@
+
+_ = get_app()._tr
+
+def makeLegalFileName(s: str):
```
2 blank lines between imports and the 1st function declaration

jonoomph commented on src/windows/exportClips.py:
```diff
@@ -0,0 +1,236 @@
+class clipExportWindow(QDialog):
+    """A popup to export clips as mp4 files
+    in a folder of the user's choosing"""
+    fp: filePicker
```
Again, something like:

```
from PyQt5.QtWidgets import QFileDialog
dialog = QFileDialog()
foo_dir = dialog.getExistingDirectory(self, 'Select an awesome directory')
```

jonoomph commented on src/windows/exportClips.py:
```diff
@@ -0,0 +1,236 @@
+        self.exporting = False
+
+    def _updateProgressBar(self, count: int, total: int):
+        d = count - total
```
Perhaps a comment here to describe this logic? Its a bit confusing to read

jonoomph commented on src/windows/exportClips.py:
```diff
@@ -0,0 +1,236 @@
+            copyFileToFolder(f, export_destination)
+            frames_written+=int(f.data.get("video_length",0))
+        for c in clips:
+            extension = "mp4"
```
This variable could be removed... and "mp4" used directly in the next line?

jonoomph commented on src/windows/exportClips.py:
```diff
@@ -0,0 +1,236 @@
+            start_time = c.data.get("start")
+            end_time = c.data.get("end")
+            fps = c.data.get("fps").get("num") / c.data.get("fps").get("den")
+            start_frame, end_frame = int(start_time * fps), int(end_time*fps)
```
Should this use round() instead of int()?

jonoomph replied:
In fact, you might move the "time to frame" calculation into a function...  The logic should be:

```
(time * fps) + 1   (to ensure we always start on frame 1)
```

jonoomph commented on src/windows/exportClips.py:
```diff
@@ -0,0 +1,236 @@
+        self._updateDialogExportFinished()
+
+    def _updateProgressBar(self, count: int, total: int):
+        if total==0:
```
Is this function duplicated? Let's try and not duplicate it.

JacksonRG replied:
Oops, I meant to move it to be next to the starting and ending export updates. The functions that disable some buttons and change the title.

jonoomph commented:
@JacksonRG Looks like a great start! Some nitpicks and small fixes. The biggest issue I have is with the custom QFileDialog we are creating... and I don't think we need to (for simplicity sake). If you have any questions, please let me know.

jonoomph commented:
@JacksonRG A few other comments:

- We should default to the home folder (or the current project folder, if any)
- Clicking "Browse" and not selecting a folder, should not clear the text box
- The progress-bar seemed incorrect to me, as it jumped to 99% immediately. Maybe the small clips were just really quickly exported... not sure, but it seemed too quick.
- Remove the "Close" button (user can always click cancel or the close icon on the window titlebar)
- Reduce the default height in the *.ui file... so there is less vertical padding between widgets (by default)
- Icon in the "Export Clips" context menu should be the same as "Export" icon
- Move "Export Clips" context option under "Split Clips...", to keep them grouped (ideally between horizontal lines), and since splitting always happens prior to exporting.

jonoomph commented:
Otherwise, this worked really nicely! Nice work!

Commit:
Improvements from review 1

Added comments explaining updateProgressBar
Made a function for time-to-frame math
Un-center the buttonbox so it looks like the export window.

JacksonRG commented:
@jonoomph The The progress bar jumps because it count's frames. So if you export a source file, as well as clips equaling the whole length, It will jump to 50% when the file is copied (half the frames) then start working on encoding the rest.

If you'd rather, the progress bar can ignore the file-copying progress.

Commit:
Suggestions from Code Review

- Right justify buttons like the export dialog
- Remove "close" button. (Covered by close button)
- Add "Done" button which hides until exporting is done
- Removed duplicate "_updateProgressBar"

Commit:
Follow file naming convention

Commit:
Cleanup Code, Improve UX

focus path textbox initially
correct typos
more precise variable names
less vertical blank space
translate text
use project directory by default
focus done button on completion
adjust window size

Commit:
Made filePicker safer. Removed completer code.

Commit:
Changed to export icon

Commit:
Remove file_picker

Commit:
Stop importing unused filePicker

jonoomph commented:
@JacksonRG Did we resolve the issues with audio-only files?  For now, it's probably fine to just default to our mp4/h.264 for any video or audio file. In the future, it might be nice to use *.mp3 files for any audio-only files.

Commit:
Export sound files as mp4. Catch Errors

All files will be exported as mp4
Make extension responsibility of `nameOfExport` function
Catch and report failed exports without crashing openShot

JacksonRG commented:
Having issues testing audio files. They keep returning True `has_video`... but causing `ffmpeg` issues when I setup the writer. So I'm adding an error popup, and catching the error.

jonoomph commented:
@JacksonRG Some files, such as MP3, do contain both video and audio. However, it should work fine to write those files as a MP4.

Commit:
Solved Audio Export Errors

Only write 'aac' encoded audio
Explain the progress bar update better
Remove files if an error occurs in their creation
    but only if the file exists
Reduce the total_frames by the frames in any clip that
    fails to export

JacksonRG commented on src/windows/export_clips.py:
```diff
@@ -101,28 +101,28 @@ def startAndEndFrames(clip):
-                               clip.data.get("audio_bit_rate")
-                               )
+    writer.SetAudioOptions(True,
+                           "aac",
```
Here was the problem! I was by default using whatever audio encoding the file had. (even when they were mp4 files this caused a crash)

jonoomph commented:
@JacksonRG LGTM! Merge away
